### PR TITLE
Fix a bug for wrong stride in fake tensor

### LINKED ISSUE #141426: Wrong output strided for example faker tensor in inductor
### 🐛 Describe the bug

# Summary
I found there may be some wrong in the faker tensor output for inductor when I was fixing trying #140399.

In the inductor, I understand pytorch seems will generate some `fake tensor`, which have the shape and stride info for the output. 
https://github.com/pytorch/pytorch/blob/3473dfa69859f4049abe46075a91014922af198c/torch/_subclasses/fake_tensor.py#L1371

If there is something I understand wrong, please correct me, thanks!

# Case study
In the below snippet:
```(python)
import torch

t = torch.zeros((5, 3, 6), dtype=torch.complex128, device='cuda')

@torch.compile
def func(a):
    print("a:")
    print("size:", a.shape)
    print("stride:", a.stride())
    b = torch.ops.aten._fft_c2c.default(a, [1], 1, True)
    print("b:")
    print("size:", b.shape)
    print("stride:", b.stride())
    c = torch.ops.aten._conj_physical.default(b)
    print("c:")
    print("size:", c.shape)
    print("stride:", c.stride())
    return c

func(t)
```
## Behavior

Its output is:
```
a:
size: torch.Size([5, 3, 6])
stride: (18, 6, 1)
b:
size: torch.Size([5, 3, 6])
stride: (18, 1, 3)
c:
size: torch.Size([5, 3, 6])
stride: (18, 1, 3)
```

**Let's focus on the stride of `c`, it's `(18, 1, 3)`, but if we draw the IRs, we could see the image:**

test-cuda-base

**We could see the stride for `aten._conj_phisical.default`(the pink one) is `(18, 6, 1)`, but not `(18, 1, 3)`.**

There is a miss-align in this case, I'm not sure whether it meet our expectation.

## Why it works now

It may generate code like:
```
def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (5, 3, 6), (18, 1, 3))
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        # Topologically Sorted Source Nodes: [buf1], Original ATen: [aten._conj_physical]
        buf0 = torch.ops.aten._conj_physical.default(arg0_1)
        del arg0_1
        buf1 = buf0
        del buf0
    return (buf1, )
```

We could see there is a line `buf0 = torch.ops.aten._conj_physical.default(arg0_1)`, but we didn't generate an `assert_size_stride` for `buf0`.

The reason why we didn't generate an `assert` here is that we may use a `MultiOutputLayout` for it.
https://github.com/pytorch/pytorch/blob/3473dfa69859f4049abe46075a91014922af198c/torch/_inductor/ir.py#L6644-L6653

# How to fix it
I'm going to draft a PR to force set the stride for the fake tensor after it was initialized, WDYT? Thanks you for your time.

### Versions

Collecting environment information...
PyTorch version: 2.6.0a0+git9f1a680
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 10.5.0-1ubuntu1~22.04) 10.5.0
Clang version: Could not collect
CMake version: version 3.31.0
Libc version: glibc-2.35

Python version: 3.11.10 (main, Oct  3 2024, 07:29:13) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-48-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.77
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           12th Gen Intel(R) Core(TM) i7-12700K
CPU family:                           6
Model:                                151
Thread(s) per core:                   2
Core(s) per socket:                   12
Socket(s):                            1
Stepping:                             2
CPU max MHz:                          5000.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7219.20
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            512 KiB (12 instances)
L1i cache:                            512 KiB (12 instances)
L2 cache:                             12 MiB (9 instances)
L3 cache:                             25 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.15.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] onnx==1.17.0
[pip3] onnxscript==0.1.0.dev20240817
[pip3] optree==0.13.1
[pip3] torch==2.6.0a0+git9f1a680
[pip3] triton==3.1.0
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] optree                    0.13.1                   pypi_0    pypi
[conda] torch                     2.6.0a0+git9f1a680           dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi
[conda] triton                    3.1.0                    pypi_0    pypi

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov

### PR DESCRIPTION
**Pull request by wenqiny**
Fixes #141426

Please see details in the issue.

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
How are the strides becoming wrong in the first place?  Can we refactor the meta kernel to use the same code as eager to set the strides of the output?

**Comment by ywq880611** [comment]
It seems the `CUFFT` lib requests to batch all dimensions not in the dim list together for output to improve performance.

Please see the code:
https://github.com/pytorch/pytorch/blob/5d7c3701e40374113921771097ebc65d9c2876bf/aten/src/ATen/native/cuda/SpectralOps.cpp#L172-L213

For example, if the output tensor's shape is `(5, 3, 6)` and stride is `(18, 6, 1)`. if we do `fft_c2c` for its second dim, its shape will be modified as  `(30, 3)` (30 = 5 * 6), and stride is `(3, 1)`.

So I suspected we may have to manually update the stride for the tensor when we back from the `CUFFT` lib invocation.

**Comment by jansel** [comment]
We should make the meta function correctly model which bakend will be chosen.

**Comment by jansel** [comment]
Based on that I think the implementation of `_prims._conj_physical_meta` is wrong. It should not call `utils.compute_elementwise_output_strides(input)`.   It is the job of the meta function to return the correct strides.

**Comment by ywq880611** [comment]
> Based on that I think the implementation of `_prims._conj_physical_meta` is wrong. It should not call `utils.compute_elementwise_output_strides(input)`. It is the job of the meta function to return the correct strides.

Uh... Sorry, I'm still not very clear about why you thought `_prims._conj_physical_meta` is wrong, because I understood that it takes an input (shape as `(5, 3, 6)`, stride as `(18, 1, 3)`), and returns a output tensor(shape as `(5, 3, 6)`, stride as `(18, 1, 3)`). Both size and stride for output is same as its input, did that not meet our expectation? Or did you mean that we would like to return a tensor with stride as `(18, 6, 1)` here, then modify the stride in some else place latter?

I'm not sure whether my understand is correct or not, but I thought the root cause may be here:
https://github.com/pytorch/pytorch/blob/4ae1c4cbb59deeae2aabc61ea1e02cf1d5992812/torch/_prims_common/wrappers.py#L315-L321
We copy the size but not stride here, hence it returns a tensor with stride as `(18, 6, 1)`.

**Comment by jansel** [comment]
> Both size and stride for output is same as its input, did that not meet our expectation?

No, both the size and the stride should be the same as *running the kernel with real tensors*.  Meta functions are required to model *both* strides and sizes accurately.  If _conj_physical returns a different stride than its input, then the meta function must do the same.

**Comment by ywq880611** [comment]
For this case with `cuda` as device (`cpu` is same):

```
import torch

t = torch.zeros((5, 3, 6), dtype=torch.complex128, device='cuda')

def func(a):
    print("a:")
    print("size:", a.shape)
    print("stride:", a.stride())
    b = torch.ops.aten._fft_c2c.default(a, [1], 1, True)
    print("b:")
    print("size:", b.shape)
    print("stride:", b.stride())
    c = torch.ops.aten._conj_physical.default(b)
    print("c:")
    print("size:", c.shape)
    print("stride:", c.stride())
    return c

func(t)
```

its output is:
```
a:
size: torch.Size([5, 3, 6])
stride: (18, 6, 1)
b:
size: torch.Size([5, 3, 6])
stride: (18, 1, 3)
c:
size: torch.Size([5, 3, 6])
stride: (18, 1, 3)
```

For the `_conj_physical` _running with a real tensor_ (stride as `(18, 1, 3)`) will also return an output tensor with stride as `(18, 1, 3)`, it was aligned with `_prims._conj_physical_meta`.

I suspected the reason why we see stride as `(18, 6, 1)` is that it was resized after calling the meta function.

**Comment by jansel** [comment]
> I suspected the reason why we see stride as `(18, 6, 1)` is that it was resized after calling the meta function.

Where is this resize coming from?

**Comment by ywq880611** [comment]
It's from here:

https://github.com/pytorch/pytorch/blob/4ae1c4cbb59deeae2aabc61ea1e02cf1d5992812/torch/_prims_common/wrappers.py#L315-L321

The `result` variable in the snippet came from the meta function, its stride is `(18, 1, 3)`, then the function will take `out` as return value, so it copy the `shape` from `result` to `out`, but not copy the `stride`, so after this method, the stride of `out` is `(18, 6, 1)` (because we just copy shape only, the `(18, 6, 1)` was inference with shape of `(5, 3, 6)`).

If you thought it's the root cause, we could fix it like this [commit](https://github.com/pytorch/pytorch/pull/141427/commits/9e62eb01c2228cb2dac4cd82b151b74a16b343e9#).

**Comment by jansel** [comment]
The example you gave doesn't seem to have an `out=` arg, so where is that out coming from?

We should change the FakeTensor behavior to match eager, rather than changing eager to match the FakeTensor (and introducing a copy).

**Comment by ywq880611** [comment]
Yes, I'm also curious about where the `out` came from, but there is a weird call stack, I couldn't found where the `out` came from in python side, so I suspect it may be from cpp side, here are some code snippets:

1) It will call `_conj_physical`
https://github.com/pytorch/pytorch/blob/2682e5e0d48a8200c1672b6a42250d3c8de44190/aten/src/ATen/native/UnaryOps.cpp#L619-L624
At this place, there is just `self` tensor.
2) then it will call `unary_op_impl`
https://github.com/pytorch/pytorch/blob/2682e5e0d48a8200c1672b6a42250d3c8de44190/aten/src/ATen/native/UnaryOps.cpp#L461-L465
We allocate a empty tensor here: `Tensor result = at::empty({0}, self.options()); `
**I suspect it's the source for the `out` came from in the python code.**

I tried to update the [PR](https://github.com/pytorch/pytorch/pull/141427/commits/2d5ed4f42e5645b501d71433dfaeca44cc08d242) to assign a tensor with same shape and stride for the `_conj_physical`, WDYT?

**Comment by jansel** [comment]
How is the out tensor in C++ making the FakeTensor code in `torch/_prims/__init__.py` incorrect?

I'll repeat, we should change the FakeTensor behavior to match the non-FakeTensor behavior.

**Comment by jansel** [comment]
> 2. Both copy size and stride from fake result tensor to out tensor, like this PR.

This one seems closer to correct, since it is fixing the FakeTensor behavior (rather than the non-FakeTensor behavior).  Though I am still confused as to where `out=` is coming from.

Can you try changing your code to:
```
c = torch.ops.aten._conj_physical(b, out=torch.empty_like(b))
```
If you actually pass a real `out=` value, does the eager mode op respect the original strides?  Or does it mutate the strides like you are doing with `as_strided_`?

Whatever we do should be correct both with and without the `out=` arg.  I worry this solution will make the `out=` version incorrect.

Also, this wrapper in wrapper.py is used in many places and this behavior might not apply to all ops.  Another solution would be to remove the wrapper from _conj_physical_meta.  This will let you define the `out=` behavior yourself to match eager mode.

**Comment by ywq880611** [comment]
I understand the reason why there is an `out` even we didn't explicitly passed it is that `conj_physical` was implement based on `conj_physical_out`, which means we must pass two arguments to `conj_physical_out`, even we didn't explicitly pass an `out` tensor to `conj_physical`, there is a code snippet for `conj_physical` (`torch/csrc/autograd/generated/python_torch_functions_2.cpp` from generated files, couldn't be linked to repo file):
```cpp
  if (_r.isNone(1)) {
    // aten::conj_physical(Tensor self) -> Tensor
    
    auto dispatch_conj_physical = [](const at::Tensor & self) -> at::Tensor {
      pybind11::gil_scoped_release no_gil;
      return self.conj_physical();
    };
    return wrap(dispatch_conj_physical(_r.tensor(0)));
  } else {
    // aten::conj_physical.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
    
    auto dispatch_conj_physical_out = [](at::Tensor out, const at::Tensor & self) -> at::Tensor {
      pybind11::gil_scoped_release no_gil;
      return at::conj_physical_out(out, self);
    };
    return wrap(dispatch_conj_physical_out(_r.tensor(1), _r.tensor(0)));
  }
```
There are two branches:
1) For the first `if (_r.isNone(1))`, which means there is no explicit `out` argument, so **it will allocate an empty tensor** like `Tensor result = at::empty({0}, self.options());` in `return self.conj_physical();` before it call to `conj_physical_out`.
2) For the else branch, it means **we explicitly pass an `out` tensor** in `return at::conj_physical_out(out, self);`, so **it didn't need to allocate an empty tensor** for it.

As your question about `c = torch.conj_physical(b, out=torch.empty_like(b))`, I tried it, it will go through the 2nd branch mentioned above, so it will return a correct result.

BTW, I update the PR, could you help me run CI again? Thank you!

**Comment by jansel** [comment]
Can you add a test for this?

I think this new version is subtly wrong.  If you pass an `out=` to `_conj_physical` it will respect the strides of the existing `out=`.

```py
>>> torch.ops.aten._conj_physical(b, out=torch.empty_like(a)).stride()
(18, 6, 1)
>>> torch.ops.aten._conj_physical(b, out=torch.empty_like(b)).stride()
(18, 1, 3)
>>> torch.ops.aten._conj_physical(b, out=torch.empty_strided((5, 3, 6), (1, 5, 15), dtype=torch.complex128, device='cuda')).stride()
(1, 5, 15)
>>> 
```
calling `out.as_strided_()` will discard the existing strides.  Which is only correct in the case that no `out=` is passed.  So it would be better to fix where the default (`out=None`) out is allocated to give that the correct strides.

**Comment by ywq880611** [comment]
## Case
Your case is interesting, but I tried it with `meta` device, it seems returns the same results, I tried to debug it and see why it works for `meta` device.
The code looks like:
```
import torch

t = torch.zeros((5, 3, 6), dtype=torch.complex128, device="meta")

def func(a):
    b = torch._fft_c2c(a, [1], 1, True)
    print(torch.ops.aten._conj_physical(b, out=torch.empty_strided((5, 3, 6), (1, 5, 15), dtype=torch.complex128, device='meta')).stride())
    print(torch.conj_physical(b, out=torch.empty_strided((5, 3, 6), (1, 5, 15), dtype=torch.complex128, device='meta')).stride())

func(t)
```

it's output is:
```
(1, 5, 15)
(18, 1, 3)
```
So we could see we got **expected result** from `torch.ops.aten._conj_physical`, but got **unexpected result** from `torch.conj_physical`, here are a simple analysis:

## torch.ops.aten._conj_physical
1) The arguments for `_fn`:
https://github.com/pytorch/pytorch/blob/4ae1c4cbb59deeae2aabc61ea1e02cf1d5992812/torch/_prims_common/wrappers.py#L270-L272
the out seems came from `Tensor result = at::empty({0}, self.options());`, so it's not the out we passed in.
1) At `_fn` it do modify the stride same as input tensor after my modification:
https://github.com/pytorch/pytorch/commit/f7afb5194747ebaaa7d78c595d5eef9afb34bb17#diff-ec5749db5ba92a707aa61e5358775eb5385a2a01ebeed49dd72c70fee979e894L320-L321
Its stride is `(18, 1, 3)` now (same as `b`).
2) After return from the `_fn` function, it seems it will go to a cpp function (in `build/aten/src/ATen/CompositeViewCopyKernels.cpp`):
```
at::Tensor & _conj_physical_out(const at::Tensor & self, at::Tensor & out) {
  auto tmp_output = at::_ops::_conj_physical::call(self);
    resize_out_helper(out, tmp_output);
  copy_arg(out, tmp_output);
  return out;
}
```
There is a `resize_out_helper(out, tmp_output);` to modify its output same as the out `out=torch.empty_strided((5, 3, 6), (1, 5, 15), dtype=torch.complex128, device='cuda')`

To be honest, I'm a little dizzy about the invocation logic between `cpp` and `python` and the logic, but it seems the `resize_out_helper(out, tmp_output);` will keep corretness.

## torch.conj_physical
it' wrong as you said.
1. The arguments for `_fn`:
https://github.com/pytorch/pytorch/blob/4ae1c4cbb59deeae2aabc61ea1e02cf1d5992812/torch/_prims_common/wrappers.py#L270-L272
the `out` have stride as `(1, 5, 15)`, so it's what we passed for the `torch.conj_physical`.
2. It didn't go to the `at::Tensor & _conj_physical_out(const at::Tensor & self, at::Tensor & out)`, so after the `as_stride_` we will get a tensor with wrong stride.

## Bug fix
I tried to fix this bug by checking whether the `out` is an empty tensor or not for `_fn` method, it works for this case, but I'm still not very clear about the mechanism diff between `torch.ops.aten._conj_physical` and `torch.conj_physical`, I guess it was caused by the different call stack and dispatch way of the two method, I will try to investigate it later.

PTAL, thanks!

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
Thanks, I think this is right now.  Though it might be cleaner to refactor this behavior into `_maybe_resize_out`.

**Comment by ywq880611** [comment]
Thank you for your review! I tried to refactor the `_maybe_resize_out`, PTAL! And I also saw there are some tests failed:
1. [test bot 1](https://github.com/pytorch/pytorch/actions/runs/12353675806/job/34513803983) failed with `Error: Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/home/ec2-user/actions-runner/_work/pytorch/pytorch/.github/actions/upload-sccache-stats'. Did you forget to run actions/checkout before running your local action?`
2. [test bot 2](https://github.com/pytorch/pytorch/actions/runs/12353675865/job/34513799198?pr=141427) failed with `Item "object" of "Tensor | _T" has no attribute "stride"`.

I have no idea about how to fix them... I suspected the first kind is not related to this PR, and the second kind is related to the change, but I have no idea about how to fix them, any insights would be very appreciated, thanks again!

**Comment by jansel** [comment]
For those tests, you may need to rebase against the `viable/strict` branch (which only gets updated when upstream tests pass)

It looks like there are some other test failures too though.

**Review by jansel — CHANGES_REQUESTED** [review]
Looks like this is failing tests.  Please re-request review when you want me to look at it again.

**Comment by ywq880611** [comment]
Hi @jansel , merry Christmas!

## Case
There is a case came in my mind, like `torch.conj_physical(b, out=torch.empty(0, dtype=torch.complex128, device='meta')`).

If we pass an explicit `out` argument with `torch.empty(0)`, it will be passed into `_fn`:
https://github.com/pytorch/pytorch/blob/4ae1c4cbb59deeae2aabc61ea1e02cf1d5992812/torch/_prims_common/wrappers.py#L270-L272

## Behavior
We will see an empty tensor for `out` in `_fn`, so we could image there is ambiguity for `out` in `_fn`, it will be either **a tensor form the explicit assignment** or **a tensor from some ops' implement**(like `conj_physical`, it was implement based on `conj_physical_out`, so it will pass an empty `out` tensor even the user didn't pass anything. 

## Solution:
Under such background, we couldn't know whether the `out` in `_fn` is an **user defined behavior** or **some other logic from the ops' implement**, then we couldn't do restride here because we may hurt correctness, there are some potential solution in my mind, like:
1. Don't wrap the ops (like `conj_physical`) with `_fn`.
2. Refactor the current implement of `conj_physical` to make it didn't rely on `conj_physical_out`.

Are these solutions feasible? WDYT? Any suggestion is appreciated! And happy new year :)

**Comment by jansel** [comment]
I'd be ok updating conj_physical to use empty_like() rather than empty({0}) and no rely on the output resise behavior.

**Comment by ywq880611** [comment]
The PR was updated, PTAL, thanks!

**Review by jansel — CHANGES_REQUESTED** [review]
Looks like you need to remove some xfails to fix CI.

**Comment by ywq880611** [comment]
It seems there is a known [issue](https://github.com/pytorch/pytorch/pull/106319), so I tried to remove the `xfail`, PTAL, thanks!

**Comment by jansel** [comment]
I approved CI, things still seem to be failing

**Comment by ywq880611** [comment]
I removed more xfail, could you please help me to run CI again, thanks!

**Review by jansel — APPROVED** [review]

File: aten/src/ATen/native/UnaryOps.cpp [hunk 1]
```diff
@@ -620,7 +620,12 @@ Tensor _conj_physical(const Tensor& self) {
   if (self.is_conj()) {
     return self.conj().clone();
   }
-  return unary_op_impl(self, at::conj_physical_out);
+  auto options = self.options();
+  if (c10::isIntegralType(self.scalar_type(), /*includeBool=*/true)) {
```
Comment (wizzniu) [inline aten/src/ATen/native/UnaryOps.cpp:624]: Sry, @ywq880611 @jansel . I am curious why change the dtype of result here(from Int to float)? Doesn't it raise error when calling `at::conj_physical_out(result, self)` in the subsequent code?

**Comment by wizzniu** [comment]
Before this PR merged, the following code can run normally:
```python
import torch
x = torch.ones((2,3,4), dtype=torch.int, device='mlu')
out_cpu = torch._conj_physical(x)
```
After this PR merged, it would raise error just like: ``RuntimeError: Found dtype Float but expected Int``.

File: aten/src/ATen/native/UnaryOps.cpp (hunk 1 above)
Comment (ywq880611) [reply aten/src/ATen/native/UnaryOps.cpp:624]: Oops, you're right, I suppose we shouldn't convert type here, I will submit a PR to fix it latter.

**Comment by ywq880611** [comment]
Hi @jansel , I'm sorry that I may make some mistake in this PR, I shouldn't convert the data type, here is the new [PR](https://github.com/pytorch/pytorch/pull/144391/files) to fix it, PTAL, thanks!

**Pull request closed without merging**